File: Data Structures & Algorithms/binary-tree-inorder-traversal/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        output = []
        dfs(root, output)
        return output
        
def dfs(root: Optional[TreeNode], output: List[int]):
    if not root:
        return;
    
    dfs(root.left, output)
    output.append(root.val)
    dfs(root.right, output)
        
        

File: Data Structures & Algorithms/binary-tree-postorder-traversal/submission-0.rs
use std::rc::Rc;
use std::cell::RefCell;

impl Solution {
    pub fn postorder_traversal(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
        let mut ans: Vec<i32> = Vec::new();

        fn dfs(root: Option<Rc<RefCell<TreeNode>>>, ans: &mut Vec<i32>) {
            if let Some(node) = root {
                let node_borrow = node.borrow();
                
                dfs(node_borrow.left.clone(), ans);
                dfs(node_borrow.right.clone(), ans);

                ans.push(node_borrow.val);
            }
        }
        dfs(root, &mut ans);
        ans
    }
}